### PR TITLE
#3783 Equality conditions with zero on voltage removed and, if necessary, replaced

### LINKED ISSUE #3783: usage of locally non differentiable functions in BaseHvdcP model (DynaFlow)
causing JacobianWithNanInf erros in some corner cases

### PR DESCRIPTION
**Pull request by ErwanGuichard49**
Every non-regression test is successful. BaseHvdcP has been preserved as it is (except for style) since the original issue stemmed from its equations. The goal of this pull request is to rely as much as possible on a boolean variable like running.value instead of a comparison of terminal.V to zero.

### CODE REVIEW

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses/BaseHvdcP.mo [hunk 1]
```diff
@@ -36,26 +36,28 @@ partial model BaseHvdcP "Base dynamic model for HVDC links with a regulation of
   parameter Types.Angle UPhase20 "Start value of voltage angle and filtered voltage angle at terminal 2 in rad";
 
 equation
+  s2Pu = Complex(P2Pu, Q2Pu);
+  s2Pu = terminal2.V * ComplexMath.conj(terminal2.i);
+
   if ((terminal1.V.re == 0) and (terminal1.V.im == 0)) then
```
Comment (joyelfeghali) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses/BaseHvdcP.mo:39]: I didn't understand why putting these equations in the if running.value loop doesn't work?

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Loads/LoadAlphaBeta.mo [hunk 2]
```diff
@@ -20,14 +20,9 @@ model LoadAlphaBeta "Load with voltage-dependent active and reactive power (alph
...
-      QPu = 0;
-    else
-      PPu = PRefPu * (1 + deltaP) * ((ComplexMath.'abs'(terminal.V) / ComplexMath.'abs'(u0Pu)) ^ alpha);
-      QPu = QRefPu * (1 + deltaQ) * ((ComplexMath.'abs'(terminal.V) / ComplexMath.'abs'(u0Pu)) ^ beta);
-    end if;
+  if running.value then
+    PPu = PRefPu * (1 + deltaP) * ((ComplexMath.'abs'(terminal.V) / ComplexMath.'abs'(u0Pu)) ^ alpha);
+    QPu = QRefPu * (1 + deltaQ) * ((ComplexMath.'abs'(terminal.V) / ComplexMath.'abs'(u0Pu)) ^ beta);
   else
     terminal.i = Complex(0);
```
Comment (joyelfeghali) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Loads/LoadAlphaBeta.mo:27]: not sure if adding PPu = 0 and QPu =0 is necessary in the else close.

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Loads/LoadAlphaBetaRestorative.mo [hunk 3]
```diff
@@ -27,23 +27,18 @@ protected
   Types.VoltageModulePu UFilteredPu(start = ComplexMath.'abs'(u0Pu)) "Bounded filtered voltage amplitude at terminal in pu (base UNom)";
 
 equation
-  if (running.value) then
-    if ((terminal.V.re == 0) and (terminal.V.im == 0)) then
-      tFilter * der(UFilteredRawPu) = - UFilteredRawPu;
-      PPu = 0;
-      QPu = 0;
```
Comment (joyelfeghali) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Loads/LoadAlphaBetaRestorative.mo]: same for all load models

**Review by joyelfeghali — CHANGES_REQUESTED** [review]

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Loads/LoadAlphaBeta.mo (hunk 2 above)
Comment (ErwanGuichard49) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Loads/LoadAlphaBeta.mo:27]: In BaseLoad.mo, we have
`SPu = Complex(PPu, QPu);`
`SPu = terminal.V * ComplexMath.conj(terminal.i);`

Thus (PPu = 0 and QPu = 0) is equivalent to (terminal.i = Complex(0) or terminal.V = Complex(0)). If running.value is false, terminal.i = Complex(0) is realistic, unlike terminal.V = Complex(0).

I would therefore keep the code as it stands.

Comment (joyelfeghali) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Loads/LoadAlphaBeta.mo:27]: I agree

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses/BaseHvdcP.mo (hunk 1 above)
Comment (ErwanGuichard49) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses/BaseHvdcP.mo:39]: If the equations are part of the if statement, the non-regression test dynawo/nrt/data/DynaFlow/HvdcPQPropDiagramPQ/HvdcPQPropDiagramPQ.jobs is modified.

In its current form, Hvdc1 is disconnected at 30 s (running.value = false), causing a small drop of U1Pu and U2Pu from 1 to around 0.995. The voltages then return gradually to 1 pu. If the equations are included in the if statement, if running.value = false, then U1Pu = U2Pu = 0 for the remainder of the simulation.

Which result is realistic ?

Comment (joyelfeghali) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses/BaseHvdcP.mo:39]: I am wondering if only one of the sides is disconnected then the voltage won't be 0, maybe that's why they didn't put in the first place in the if running.value condition.
What do you think @gaouizer ?

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/IEC/IEC61400/BaseControls/Auxiliaries/Measurements.mo [hunk 4]
```diff
@@ -52,10 +52,6 @@ model Measurements "Measurement module for wind turbine controls (IEC N°61400-2
   Modelica.Blocks.Interfaces.RealOutput UPu(start = U0Pu) "Voltage amplitude at grid terminal in pu (base UNom)" annotation(
     Placement(visible = true, transformation(origin = {150, 0}, extent = {{-10, -10}, {10, 10}}, rotation = 0), iconTransformation(origin = {110, 0}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
 
-  //Other variables
-  Types.CurrentModulePu IWtPu(start = ComplexMath.'abs'(i0Pu) * SystemBase.SnRef / SNom) "Current module at grid terminal in pu (base UNom, SNom)";
-  Types.VoltageModulePu UWtPu(start = U0Pu) "Voltage amplitude at grid terminal in pu (base UNom)";
```
Comment (joyelfeghali) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/IEC/IEC61400/BaseControls/Auxiliaries/Measurements.mo:57]: why these were removed?

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Utilities/Measurements.mo [hunk 5]
```diff
@@ -44,18 +44,22 @@ model Measurements "This block measures the voltage, current, active power and r
...
   terminal1.i = iPu;
   terminal1.V = uPu;
-  IPhase = ComplexMath.arg(iPu);
-  IPu = ComplexMath.'abs'(iPu);
   PPu = ComplexMath.real(uPu * ComplexMath.conj(iPu));
   QPu = ComplexMath.imag(uPu * ComplexMath.conj(iPu));
-  UPhase = ComplexMath.arg(uPu);
-  UPu = ComplexMath.'abs'(uPu);
+  complexToPolarI.u = iPu;
+  IPu = complexToPolarI.len;
```
Comment (joyelfeghali) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Utilities/Measurements.mo]: this is to avoid using 'abs'?

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Shunts/ShuntB.mo [hunk 6]
```diff
@@ -32,17 +32,14 @@ model ShuntB "Shunt element with constant susceptance, reactive power depends on
...
-  else
-    UPu = ComplexMath.'abs'(terminal.V);
-  end if;
 
-  if (running.value) then
-    QPu = BPu * UPu^2;
+  if running.value then
+    QPu = BPu * UPu ^ 2;
     PPu = 0;
+    UPu = ComplexMath.'abs'(terminal.V);
```
Comment (joyelfeghali) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Shunts/ShuntB.mo:40]: Could you put back UPu equation outside of the running.value?

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Shunts/ShuntBWithSections.mo [hunk 7]
```diff
@@ -37,19 +37,16 @@ model ShuntBWithSections "Shunt element with voltage-dependent reactive power an
...
-  end if;
   SPu = Complex(PPu, QPu);
   SPu = terminal.V * ComplexMath.conj(terminal.i);
 
-  if (running.value) then
-    QPu = - BPu * UPu ^ 2;
+  if running.value then
+    QPu = -BPu * UPu ^ 2;
     PPu = 0;
+    UPu = ComplexMath.'abs'(terminal.V);
```
Comment (joyelfeghali) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Shunts/ShuntBWithSections.mo:47]: same

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Sources/InjectorBG.mo [hunk 8]
```diff
@@ -46,16 +46,18 @@ protected
   Types.ComplexAdmittancePu YPuSnRef(re = GPuSnRef, im = BPuSnRef) "Admittance in pu (base SnRef)";
 
 equation
-  UPu = ComplexMath.'abs'(terminal.V);
   PInjPu = -ComplexMath.real(terminal.V * ComplexMath.conj(terminal.i));
   QInjPu = -ComplexMath.imag(terminal.V * ComplexMath.conj(terminal.i));
 
   if running.value then
     terminal.i = terminal.V * YPuSnRef;
+    UPu = ComplexMath.'abs'(terminal.V);
```
Comment (joyelfeghali) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Sources/InjectorBG.mo]: same

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Sources/InjectorIDQ.mo [hunk 9]
```diff
@@ -65,8 +64,10 @@ equation
     // Park's transformations dq-currents in generator convention, -> receptor convention for terminal
     terminal.i.re = -(cos(UPhase) * idPu - sin(UPhase) * iqPu) * SNom / SystemBase.SnRef;
     terminal.i.im = -(sin(UPhase) * idPu + cos(UPhase) * iqPu) * SNom / SystemBase.SnRef;
+    UPu = ComplexMath.'abs'(terminal.V);
```
Comment (joyelfeghali) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Sources/InjectorIDQ.mo]: same

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/StaticVarCompensators/BaseClasses/BaseSVarC.mo [hunk 10]
```diff
@@ -55,12 +55,8 @@ equation
...
   end when;
 
-  if (running.value) then
-    if ((terminal.V.re == 0) and (terminal.V.im == 0)) then
-      UPu = 0;
-    else
-      UPu = Modelica.ComplexMath.'abs'(terminal.V);
-    end if;
+  if running.value then
+    UPu = Modelica.ComplexMath.'abs'(terminal.V);
```
Comment (joyelfeghali) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/StaticVarCompensators/BaseClasses/BaseSVarC.mo:59]: same

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses/BaseHvdcP.mo (hunk 1 above)
Comment (joyelfeghali) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses/BaseHvdcP.mo:39]: I think no need to put the UPu equation inside the if running.value, I think it's even better to keep it outside this condition or else UPu won't be coherent with terminal.V
I think when we disconnect the terminal.i go to 0, this will be enough. If we also want the voltage to change I think we should put terminal.V = 0 (but this is not the purpose of the dev, the purpose is to be safe from 'abs' when terminal.V goes to 0).

**Review by joyelfeghali — CHANGES_REQUESTED** [review]

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/IEC/IEC61400/BaseControls/Auxiliaries/Measurements.mo (hunk 4 above)
Comment (ErwanGuichard49) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/IEC/IEC61400/BaseControls/Auxiliaries/Measurements.mo:57]: UWtPu is redundant as UPu is the module of uPu too. IWtPu is the non-filtered module of iPu (the filtered module is IFiltPu), I am restoring it.

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Utilities/Measurements.mo (hunk 5 above)
Comment (ErwanGuichard49) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Utilities/Measurements.mo]: Yes.

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Loads/LoadAlphaBetaRestorative.mo [hunk 11]
```diff
@@ -28,24 +28,17 @@ protected
 
 equation
   if running then
-    if (terminal.V == Complex(0)) then
-      tFilter * der(UFilteredRawPu) = -UFilteredRawPu;
```
Comment (joyelfeghali) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Loads/LoadAlphaBetaRestorative.mo:32]: This equation was added in order to be able to reconnect the load after it is disconnected. See https://github.com/dynawo/dynawo/commit/52f7faed5b43039446767f4ea6217b339b769b51#diff-fd84e9cc6e2d9596c509ff855318c99bf6d804d8292f995492b529ef26619326
I think it would be needed or rewritten in a way.
Could you check for all these models in this commit if the related nrt file is working after removing the condition on terminal.V?

Is there a different way to implement it?

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Transformers/TransformersFixedTap/TransformerFixedRatio.mo [hunk 12]
```diff
@@ -63,23 +63,16 @@ equation
   Q2Pu = ComplexMath.imag(terminal2.V * ComplexMath.conj(terminal2.i));
 
   if running then
-    if ((terminal1.V.re == 0) and (terminal1.V.im == 0)) then
-      U1Pu = 0;
-    else
-      U1Pu = ComplexMath.'abs'(terminal1.V);
-    end if;
-    if ((terminal2.V.re == 0) and (terminal2.V.im == 0)) then
```
Comment (joyelfeghali) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Transformers/TransformersFixedTap/TransformerFixedRatio.mo:71]: are there any other models that uses if ((terminal.V.re == 0) and (terminal.V.im == 0)) or if terminal.V == Complex(0) that are not included in this dev?

**Review by joyelfeghali — CHANGES_REQUESTED** [review]
In general I think this is OK.  we just have to check for DynawoRT simulator if there are other requirements, besides the load connection after being disconnected that needs to be taken into account.

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Transformers/TransformersFixedTap/TransformerFixedRatio.mo (hunk 12 above)
Comment (ErwanGuichard49) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Transformers/TransformersFixedTap/TransformerFixedRatio.mo:71]: Every instance of these equalities has been removed.

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Loads/LoadAlphaBetaRestorative.mo (hunk 11 above)
Comment (ErwanGuichard49) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Loads/LoadAlphaBetaRestorative.mo:32]: The case nrt/data/Network/Load/Modelica/NodeBreakerLoadReconnection/DisconnectLoad.jobs fails with this new version.

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Transformers/TransformersFixedTap/TransformerFixedRatio.mo (hunk 12 above)
Comment (ErwanGuichard49) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Transformers/TransformersFixedTap/TransformerFixedRatio.mo:71]: I have removed the occurrences of "terminal.V <> 0" as well.

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Loads/LoadAlphaBetaRestorative.mo (hunk 11 above)
Comment (ErwanGuichard49) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Loads/LoadAlphaBetaRestorative.mo:32]: The case nrt/data/Network/Load/Modelica/NodeBreakerLoadReconnection/DisconnectLoad.jobs fails without the condition on terminal.V.

Comment (joyelfeghali) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Loads/LoadAlphaBetaRestorative.mo:32]: Is there a different way to implement it?
like adding UFilteredRawPu equation if if running = false?
or with another bool condition?

The last resort would be to check if instead the inequality works as well (terminal.V <= Complex(eps) and terminal.V >= Complex(eps))

Comment (ErwanGuichard49) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Loads/LoadAlphaBetaRestorative.mo:32]: With the UFilteredRawPu equation in the branch running = false, the simulation still fails. With the inequality-based condition in BaseLoad, it fails too. I am investigating this issue.

Comment (ErwanGuichard49) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Loads/LoadAlphaBetaRestorative.mo:32]: The only way the simulation succeeds is by using an inequality-based condition in BaseLoad as well as in other load models. The curves are not impacted but the .txt and .log are.

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Loads/LoadAlphaBeta.mo [hunk 13]
```diff
@@ -20,7 +20,7 @@ model LoadAlphaBeta "Load with voltage-dependent active and reactive power (alph
   parameter Real beta "Reactive load sensitivity to voltage";
 
 equation
-  if running and terminal.V <> Complex(0) then
+  if running and nonZeroVoltage then
```
Comment (joyelfeghali) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Loads/LoadAlphaBeta.mo]: Do we need this nonZeroVoltage condition everywhere in the load?

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Loads/LoadAlphaBetaRestorative.mo [hunk 14]
```diff
@@ -28,7 +28,7 @@ protected
 
 equation
   if running then
-    if (terminal.V == Complex(0)) then
+    if not(nonZeroVoltage) then
       tFilter * der(UFilteredRawPu) = -UFilteredRawPu;
       terminal.i = Complex(0);
     elseif UFilteredPu == 0 then
```
Comment (joyelfeghali) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Loads/LoadAlphaBetaRestorative.mo]: this condition is also necessary ?

**Review by joyelfeghali — CHANGES_REQUESTED** [review]

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Loads/LoadAlphaBetaRestorative.mo (hunk 11 above)
Comment (joyelfeghali) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Loads/LoadAlphaBetaRestorative.mo:32]: the inequality condition is it needed when there is no derivative on V as well?

**Review by joyelfeghali — CHANGES_REQUESTED** [review]
To be rebased and check nrt

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Loads/LoadAlphaBetaRestorative.mo [hunk 15]
```diff
@@ -28,7 +28,7 @@ protected
 
 equation
   if running then
-    if (terminal.V == Complex(0)) then
+    if not(nonZeroVoltage) then
```
Comment (rosiereflo) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Loads/LoadAlphaBetaRestorative.mo:31]: this will generate a jacobian reconstruction for the solver as it is a boolean condition, not sure it is a good thing. I would prefer to add a proper function comparing complex or reals to zero and calls it in the if conditions

Comment (ErwanGuichard49) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Loads/LoadAlphaBetaRestorative.mo]: I will deal with the load models in a separate PR.

File: nrt/data/IEEE14/IEEE14_SyntaxExamples/IEEE14_DumpFinalValues/reference/outputs/finalValues/dumpFinalValues-_LOAD__12_EC.txt [hunk 16]
```diff
@@ -11,9 +11,9 @@ load_terminal_i_im                                : y =      -0.029685 yp =
 load_terminal_i_re                                : y =       0.051880 yp =      -0.000000
 load_SPu_im                                       : alias of load_QPu
 load_SPu_re                                       : alias of load_PPu
- ====== INIT CALCULATED VARIABLES VALUES ======
+ ====== CALCULATED VARIABLES VALUES ======
 load_UPu                                          : y =       1.055412
- ====== INIT DISCRETE VARIABLES VALUES ======
```
Comment (joyelfeghali) [inline nrt/data/IEEE14/IEEE14_SyntaxExamples/IEEE14_DumpFinalValues/reference/outputs/finalValues/dumpFinalValues-_LOAD__12_EC.txt:16]: All nrt are OK?

**Review by joyelfeghali — CHANGES_REQUESTED** [review]

**Review by joyelfeghali — DISMISSED** [review]

File: nrt/data/IEEE14/IEEE14_SyntaxExamples/IEEE14_DumpFinalValues/reference/outputs/finalValues/dumpFinalValues-_LOAD__12_EC.txt (hunk 16 above)
Comment (ErwanGuichard49) [reply nrt/data/IEEE14/IEEE14_SyntaxExamples/IEEE14_DumpFinalValues/reference/outputs/finalValues/dumpFinalValues-_LOAD__12_EC.txt:16]: Yes

**Review by rosiereflo — CHANGES_REQUESTED** [review]
to be able to apply the same epsilon on OpenModelica and Dynawo it would be better to add share functions that will be later on overriden on dynawo side. Moreover using booleans in if conditions has side effects on the number of Jacobian computed

**Comment by ErwanGuichard49** [comment]
> to be able to apply the same epsilon on OpenModelica and Dynawo it would be better to add share functions that will be later on overriden on dynawo side. Moreover using booleans in if conditions has side effects on the number of Jacobian computed

eps is used in dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses/BaseHvdcP.mo only. @joyelfeghali should I revert the changes pertaining to this model ?

**Comment by joyelfeghali** [comment]
> > to be able to apply the same epsilon on OpenModelica and Dynawo it would be better to add share functions that will be later on overriden on dynawo side. Moreover using booleans in if conditions has side effects on the number of Jacobian computed
> 
> eps is used in dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses/BaseHvdcP.mo only. @joyelfeghali should I revert the changes pertaining to this model ?

I suggest keeping in this branch the models where we only remove the terminal.V==0 without replacing it with something else.
The models that still need these conditions can be studied in a specified branch, where we can focus on the impact of the boolean/eps on Dynawo side.

**Comment by rosiereflo** [comment]
> > > to be able to apply the same epsilon on OpenModelica and Dynawo it would be better to add share functions that will be later on overriden on dynawo side. Moreover using booleans in if conditions has side effects on the number of Jacobian computed
> > 
> > 
> > eps is used in dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses/BaseHvdcP.mo only. @joyelfeghali should I revert the changes pertaining to this model ?
> 
> I suggest keeping in this branch the models where we only remove the terminal.V==0 without replacing it with something else. The models that still need these conditions can be studied in a specified branch, where we can focus on the impact of the boolean/eps on Dynawo side.

Sounds good, are we sure this does not create regressions during connection scenarios?

**Comment by joyelfeghali** [comment]
> > > > to be able to apply the same epsilon on OpenModelica and Dynawo it would be better to add share functions that will be later on overriden on dynawo side. Moreover using booleans in if conditions has side effects on the number of Jacobian computed
> > > 
> > > 
> > > eps is used in dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses/BaseHvdcP.mo only. @joyelfeghali should I revert the changes pertaining to this model ?
> > 
> > 
> > I suggest keeping in this branch the models where we only remove the terminal.V==0 without replacing it with something else. The models that still need these conditions can be studied in a specified branch, where we can focus on the impact of the boolean/eps on Dynawo side.
> 
> Sounds good, are we sure this does not create regressions during connection scenarios?

I think it was mainly related to loads but @vermeulenthi can confirm?

**Comment by ErwanGuichard49** [comment]
I have just reversed the changes related to BaseHvdcP and restored a nrt reference.

**Comment by vermeulenthi** [comment]
> > > > > to be able to apply the same epsilon on OpenModelica and Dynawo it would be better to add share functions that will be later on overriden on dynawo side. Moreover using booleans in if conditions has side effects on the number of Jacobian computed
> > > > 
> > > > 
> > > > eps is used in dynawo/sources/Models/Modelica/Dynawo/Electrical/HVDC/BaseClasses/BaseHvdcP.mo only. @joyelfeghali should I revert the changes pertaining to this model ?
> > > 
> > > 
> > > I suggest keeping in this branch the models where we only remove the terminal.V==0 without replacing it with something else. The models that still need these conditions can be studied in a specified branch, where we can focus on the impact of the boolean/eps on Dynawo side.
> > 
> > 
> > Sounds good, are we sure this does not create regressions during connection scenarios?
> 
> I think it was mainly related to loads but @vermeulenthi can confirm?

The NRT NodeBreakerLoadReconnection covers the reconnection of Load models. For other models, having them fail reconnection would be a good opportunity to add new NRT (and fix), but I don't see risky modifications in this branch.

**Review by joyelfeghali — APPROVED** [review]
